### PR TITLE
Make Grid and and Graph view work with task mapping

### PR DESCRIPTION
**Pull request by bbovenzi**
Problem:
- UI crashes when trying to render a DAG with task mapping

Solution:
- Add a new check to prevent crashing
- Add a "summary" to the tooltip of a task instance that is mapped
- Add `[ ]` to the label of tasks that are mapped. The count of tasks is included in the graph view and updates with autorefresh

Screen Shot 2022-02-22 at 12 22 54 PM

Screen Shot 2022-02-22 at 12 22 31 PM

---

**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/www/static/js/tree/InstanceTooltip.jsx [hunk 1]
```diff
@@ -24,30 +24,33 @@ import { Box, Text } from '@chakra-ui/react';
...
+  ['upstream_failed', 0],
+  ['up_for_retry', 0],
+  ['up_for_reschedule', 0],
+  ['running', 0],
+  ['deferred', 0],
+  ['sensing', 0],
+  ['queued', 0],
+  ['scheduled', 0],
+  ['skipped', 0],
+  ['no_status', 0],
```
Comment (ashb) [inline airflow/www/static/js/tree/InstanceTooltip.jsx]: Not thrilled about having this list in multiple JS files. Is there any sensible way to define it once and reuse it?

File: airflow/www/utils.py [hunk 2]
```diff
@@ -53,10 +56,82 @@ def datetime_to_string(value: Optional[DateTime]) -> Optional[str]:
...
+        'deferred',
+        'sensing',
+        'running',
+        'shutdown',
+        'restarting',
+        'removed',
+        'no_status',
+        'success',
+        'skipped',
+    ]
```
Comment (ashb) [inline airflow/www/utils.py]: These should probably be TaskInstanceState.FAILED etc.

(which when treated like strings will give you the same value you have, with the exception of None/"no_status".

Comment (bbovenzi) [reply airflow/www/utils.py]: Updated.

File: airflow/www/static/js/tree/InstanceTooltip.jsx (hunk 1 above)
Comment (bbovenzi) [reply airflow/www/static/js/tree/InstanceTooltip.jsx]: Made a `utils.js` file for this and other similar uses.

File: airflow/www/static/js/tree/InstanceTooltip.jsx [hunk 3]
```diff
@@ -88,6 +97,17 @@ const InstanceTooltip = ({
           {groupSummary}
         </>
       )}
+      {group.isMapped && (
+        <>
+          <br />
+          <Text as="strong">
+            {mappedStates.length}
+            {' '}
+            Tasks Mapped
```
Comment (ryanahamilton) [inline airflow/www/static/js/tree/InstanceTooltip.jsx]: Should this be `mappedStates.length === 1 ? 'Task' : 'Tasks'`?

**Review by ryanahamilton — APPROVED** [review]
I'm probably not the best to review the python changes, but otherwise LGTM

**Comment by uranusjr** [comment]
What do *Startd*, *Ended* and *Duration* mean in this context? Should we change those as well?

**Comment by bbovenzi** [comment]
> What do _Startd_, _Ended_ and _Duration_ mean in this context? Should we change those as well?

They are the overall start, end, duration. From the map_index 0 to the last map index. Just like how it is calculated for a group

File: airflow/www/static/js/tree/InstanceTooltip.jsx (hunk 3 above)
Comment (bbovenzi) [reply airflow/www/static/js/tree/InstanceTooltip.jsx]: Updated!

**Comment by ashb** [comment]
Are the screenshots in the PR description up-to-date?

**Comment by bbovenzi** [comment]
> Are the screenshots in the PR description up-to-date?

Yes. I have some more advanced changes in a later PR.

File: airflow/www/utils.py [hunk 4]
```diff
@@ -53,10 +56,82 @@ def datetime_to_string(value: Optional[DateTime]) -> Optional[str]:
...
+        TaskInstanceState.UP_FOR_RESCHEDULE,
+        TaskInstanceState.QUEUED,
+        TaskInstanceState.SCHEDULED,
+        TaskInstanceState.DEFERRED,
+        TaskInstanceState.SENSING,
+        TaskInstanceState.RUNNING,
+        TaskInstanceState.SHUTDOWN,
+        TaskInstanceState.RESTARTING,
+        TaskInstanceState.REMOVED,
+        'no_status',
```
Comment (ashb) [inline airflow/www/utils.py]: I'm confused how this ever worked -- I can't see how a TI in state None gets mapped to this no_status grouping.

Comment (ashb) [reply airflow/www/utils.py]: ("this ever worked" being the existing `get_summary` function)

Comment (uranusjr) [reply airflow/www/utils.py]: It probably never did and `group_state` simply is never `no_status`?

Comment (uranusjr) [reply airflow/www/utils.py]: I’d say lets remove `no_status` here (and from `get_summary`). But not in js code—it seems like this is the initial UI state.

Comment (bbovenzi) [reply airflow/www/utils.py]: Good call. I removed `no_status` and updated `get_summary` to also use `TaskInstanceState` instead of strings

**Review by uranusjr — APPROVED** [review]

File: airflow/www/views.py [hunk 5]
```diff
@@ -223,41 +223,47 @@ def get_date_time_num_runs_dag_runs_form_data(www_request, session, dag):
...
         return {
             'id': task_item_or_group.task_id,
-            'instances': [wwwutils.encode_ti(ti) for ti in tis if ti.task_id == task_item_or_group.task_id],
+            'instances': [
+                wwwutils.encode_ti(ti, task_item_or_group.is_mapped, session)
+                for ti in tis
+                if ti.task_id == task_item_or_group.task_id
+            ],
             'label': task_item_or_group.label,
-            'extra_links': task_item_or_group.extra_links,
```
Comment (wojsamjan) [inline airflow/www/views.py:236]: @bbovenzi 
Why do we change: 
'extra_links': task_item_or_group.extra_links,
to
'extra_links': [],

This change stands for not showing links in Task status in Grid UI

Comment (bbovenzi) [reply airflow/www/views.py:236]: Good catch! That was supposed to be temporary during development. Fix: https://github.com/apache/airflow/pull/21971

**Pull request merged**